setup.py

```diff
@@ -0,0 +1,190 @@
+        else:
+            import peigen
+
+            return peigen.header_path
+
+
+ext_modules = [
+    Extension(
+        "lsst.scarlet.lite.operators_pybind11",
+        [os.path.join(package_path, "operators_pybind11.cc")],
+        include_dirs=[
+            GetPybindInclude(),
+            GetPybindInclude(user=True),
+            GetEigenInclude(),
+        ],
+        language="c++",
+    ),
+    Extension(
+        "lsst.scarlet.lite.detect_pybind11",
+        [os.path.join(package_path, "detect_pybind11.cc")],
+        include_dirs=[
+            GetPybindInclude(),
+            GetPybindInclude(user=True),
+            GetEigenInclude(),
+        ],
+        language="c++",
+    ),
+]
+
+
+# As of Python 3.6, CCompiler has a `has_flag` method.
+# cf http://bugs.python.org/issue26689
+def has_flag(compiler, flagname):
+    """Return a boolean indicating whether a flag name is supported on
+    the specified compiler.
+    """
+    import tempfile
+
+    with tempfile.NamedTemporaryFile("w", suffix=".cpp") as f:
+        f.write("int main (int argc, char **argv) { return 0; }")
+        try:
+            compiler.compile([f.name], extra_postargs=[flagname])
+        except setuptools.distutils.errors.CompileError:
+            return False
+    return True
+
+
+def cpp_flag(compiler):
+    """Return the -std=c++[11/14] compiler flag.
+    The c++14 is prefered over c++11 (when it is available).
+    """
+    if has_flag(compiler, "-std=c++14"):
+        return "-std=c++14"
+    elif has_flag(compiler, "-std=c++11"):
+        return "-std=c++11"
+    else:
+        raise RuntimeError(
+            "Unsupported compiler -- at least C++11 support " "is needed!"
+        )
+
+
+class BuildExt(build_ext):
+    """A custom build extension for adding compiler-specific options."""
+
+    c_opts = {
+        "msvc": ["/EHsc"],
+        "unix": [],
+    }
+
+    if sys.platform == "darwin":
+        c_opts["unix"] += ["-stdlib=libc++", "-mmacosx-version-min=10.7"]
+
+    def build_extensions(self):
+        ct = self.compiler.compiler_type
+        opts = self.c_opts.get(ct, [])
+        if ct == "unix":
+            opts.append('-DVERSION_INFO="%s"' % self.distribution.get_version())
+            opts.append(cpp_flag(self.compiler))
+            if has_flag(self.compiler, "-fvisibility=hidden"):
+                opts.append("-fvisibility=hidden")
+        elif ct == "msvc":
+            opts.append('/DVERSION_INFO=\\"%s\\"' % self.distribution.get_version())
+        for ext in self.extensions:
+            ext.extra_compile_args = opts
+        build_ext.build_extensions(self)
+
+
+install_requires = ["numpy>=1.17", "scipy", "pydantic"]
+# Only require the pybind11 and peigen packages if
+# the C++ headers are not already installed
+if pybind11_path is None:
+    install_requires.append("pybind11>=2.2")
+if eigen_path is None:
+    install_requires.append("peigen>=0.0.9")
+
+print(f"\n\ninstall requires: {install_requires} \n\n")
+
+setup(
```

Comment:
Most of this should not be in the setup.py but in a static configuration file -- it makes it easier for other tooling to extract the information without building the package. As discussed on slack, pyproject.toml is preferred.

Reply:
I think the pybind11 part isn't necessary anymore. I feel like there was a time when there was an incompatibility between the current release of pybind11 and the one included in the stack, but this _probably_ became obsolete once we added pybind11 to the conda environment.

Reply:
But I think that I still need to include the `install_requires` in `setup.py` because if the EIGEN headers aren't installed, pip needs to install a python package that we created just to have a wrapper for the EIGEN headers.

Reply:
Although looking at pybind11's page now, it looks like those headers are included and we might be able to remove that dependency now too.

Reply:
install_requires should not be used in setup.py to specify a build pre-req. You are meant to add it to pyproject.toml in the build-requires section. See https://github.com/lsst/sphgeom/blob/main/pyproject.toml#L1-L8

This way pip sets that all up before it even starts worrying about running your setup.py.

Reply:
So here's the problem. In the stack we include the Eigen headers with the `eigen` dependency. But for 3rd party pip installs, the `peigen` package is required _if_ the user doesn't have a local install of Eigen already. Is there a way to move that logic from `setup.py` into the `pyproject.toml`? If not, I don't see how I can get around having these in the `setup.py`.

Reply:
Optional build dependencies are not a great fit for static config files. You wrote peigen? Can it notice that eigen is already installed and do the right thing?

Reply:
Right, that's why I was using the `install_requires` in setup.py. It checked to see if there was a local `EIGEN_DIR` variable setup in the environment (for example when the science pipelines are installed) and if it's missing, then and only then would it install peigen, which is basically just a method to get the header directory.

Reply:
I suppose an alternative would just be to include the Eigen headers in scarlet lite, and only use them if there is no environment variable already installed. Does that sound like an acceptable solution?

Reply:
Please don’t vendor eigen.

Reply:
Aren't you using scons to build in a LSST environment? When you use pip you are not. Will EIGEN_DIR even be set for pip users? Why can't pip install peigen anyhow? setup.py doesn't have to import anything from it if it finds EIGEN elsewhere.

Reply:
Also, build-requires do not get installed into your standard python. pip installs them into a temporary location used just for the build itself.

Reply:
The idea was that Eigen versions are not always compatible with one another, so allowing users to set an Eigen directory would allow them to build scarlet (lite) with their preferred version of Eigen. In this case it probably isn't that big of a deal because we don't expect users to inherit from scarlet c++ primitives, so it's probably ok to just use a version of Eigen that we know works (hence peigen). So if I make peigen a required dependency, scons won't look at any of the setup files and the stack won't try to install it? If that's the case, then that does seem like the best solution.

Reply:
scons is not going to look at the pyproject.toml or the setup.py. The build-system.requires is only going to be looked at by pip and then it won't even install them in a place that anything other than that build will see. Your setup.py still can decide whether the peigen code is even imported.